README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
LATEerror
=========

**LATEerror** package implements the generalized method of moments estimation for the local average treatment effect when the treatment variable may be mismeasured. The estimation procedure is developed by Yanagi (2017).

Intsallation
============

Run the following code to install **LATEerror**.

``` r
install.packages("devtools")
devtools::install_github("tkhdyanagi/LATEerror")
library(LATEerror)
```

Usage
=====

``` r
LATEerror(Y, D, Z, V, weight = NULL, optimal = TRUE, equal = TRUE, lower = NULL, upper = NULL, controlDE = NULL, controlBFGS = NULL)
```

See also
========

1.  See package vignettes "How to use `LATEerror`" for more details.
"# LATEerror" 


NEWS.md
# LATEerror 0.1.0

* Released the first version.

# LATEerror 0.2.0

* Revised the main function and added some functions.


tests/testthat.R
library(testthat)
library(LATEerror)
library(dplyr)

test_check("LATEerror")


tests/testthat/test_errors.R

context("Testing errors")

# simulated data
n <- 1000
set.seed(102)
mydata <- data.frame(U1 = rnorm(n, 0, 1), U2 = rnorm(n, 0, 1))
mydata <- mydata %>%
  dplyr::mutate(Z = rbinom(n, size = 1, prob = 0.5)) %>%
  dplyr::mutate(V = rbinom(n, size = 1, prob = 0.5) + 1) %>%
  dplyr::mutate(Ds = ifelse(( -1.25 + Z + 0.5 * V - U1 >= 0), 1, 0)) %>%
  dplyr::mutate(D = Ds * rbinom(n, size = 1, prob = 0.8) + (1 - Ds) * rbinom(n, size = 1, prob = 0.2)) %>%
  dplyr::mutate(Y = 1 + Ds + U2)

# errors for data
expect_error(LATEerror(Y = NA, D = mydata$D, Z = mydata$Z, V = mydata$V, weight = NULL, optimal = TRUE, lower = NULL, upper = NULL, control = NULL), "The lengths of Y, D, Z, and V must be the same.")

expect_error(LATEerror(Y = mydata$Y, D = NULL, Z = mydata$Z, V = mydata$V, weight = NULL, optimal = TRUE, lower = NULL, upper = NULL, control = NULL), "The lengths of Y, D, Z, and V must be the same.")

expect_error(LATEerror(Y = mydata$Y, D = mydata$D, Z = 1, V = mydata$V, weight = NULL, optimal = TRUE, lower = NULL, upper = NULL, control = NULL), "The lengths of Y, D, Z, and V must be the same.")

expect_error(LATEerror(Y = mydata$Y, D = mydata$D, Z = mydata$Z, V = Inf, weight = NULL, optimal = TRUE, lower = NULL, upper = NULL, control = NULL), "The lengths of Y, D, Z, and V must be the same.")

# errors for weight matrix
expect_error(LATEerror(Y = mydata$Y, D = mydata$D, Z = mydata$Z, V = mydata$V, weight = 1, optimal = TRUE, lower = NULL, upper = NULL, control = NULL), "'weight' must be NULL or a symmetric matrix.")

expect_error(LATEerror(Y = mydata$Y, D = mydata$D, Z = mydata$Z, V = mydata$V, weight = diag(2), optimal = TRUE, lower = NULL, upper = NULL, control = NULL), "'weight' must be NULL or a symmetric matrix.")

expect_error(LATEerror(Y = mydata$Y, D = mydata$D, Z = mydata$Z, V = mydata$V, weight = FALSE, optimal = TRUE, lower = NULL, upper = NULL, control = NULL), "'weight' must be NULL or a symmetric matrix.")

# errors for optimal
expect_error(LATEerror(Y = mydata$Y, D = mydata$D, Z = mydata$Z, V = mydata$V, weight = NULL, optimal = 1, lower = NULL, upper = NULL, control = NULL), "'optimal' must be logical.")

expect_error(LATEerror(Y = mydata$Y, D = mydata$D, Z = mydata$Z, V = mydata$V, weight = NULL, optimal = diag(9), lower = NULL, upper = NULL, control = NULL), "'optimal' must be logical.")

# errors for lower and upper
expect_error(LATEerror(Y = mydata$Y, D = mydata$D, Z = mydata$Z, V = mydata$V, weight = NULL, optimal = TRUE, lower = 1, upper = NULL, control = NULL), "The lengths of 'lower' and 'upper' must be identical to that of the parameters.")

expect_error(LATEerror(Y = mydata$Y, D = mydata$D, Z = mydata$Z, V = mydata$V, weight = NULL, optimal = TRUE, lower = NULL, upper = NA, control = NULL), "The lengths of 'lower' and 'upper' must be identical to that of the parameters.")

expect_error(LATEerror(Y = mydata$Y, D = mydata$D, Z = mydata$Z, V = mydata$V, weight = NULL, optimal = TRUE, lower = NULL, upper = Inf, control = NULL), "The lengths of 'lower' and 'upper' must be identical to that of the parameters.")

# errors for control
expect_error(LATEerror(Y = mydata$Y, D = mydata$D, Z = mydata$Z, V = mydata$V, weight = NULL, optimal = TRUE, lower = NULL, upper = NULL, control = TRUE), "'control' must be correctly specified.")

expect_error(LATEerror(Y = mydata$Y, D = mydata$D, Z = mydata$Z, V = mydata$V, weight = NULL, optimal = TRUE, lower = NULL, upper = NULL, control = 1), "'control' must be correctly specified.")



R/moments.R
#' Computing the GMM criterion.
#'
#' \code{criterion} is an auxiliary function to implement the GMM inference on your own.
#' It returns a value of the GMM criterion.
#'
#' @param par vector of parameter values
#' @param data dataframe that contains Y, D, Z, and V with indicators
#' @param weight weighting matrix for GMM estimation
#' @param equal logical whether the misclassification probabilities do not depend on instrument
#'
#' @export
#'
criterion <- function(par, data, weight, equal) {

  # the means of the moments
  S <- mean_moments(par = par, data = data, equal = equal)

  # criterion
  return(t(S) %*% weight %*% S)

}



#' Computing the means of the moments for the GMM estimation.
#'
#' @param par vector of parameter values
#' @param data dataframe containing the outcome, treatment, instrument, exogenous variable
#' @param equal logical whether the misclassification probabilities do not depend on instrument
#'
mean_moments <- function(par, data, equal) {

  # the means of the moments
  result <- colMeans(moments(par = par, data = data, equal = equal))

  return(result)

}



#' Computing the values of the moments for the GMM estimation.
#'
#' @param par vector of parameter values
#' @param data dataframe containing the outcome, treatment, instrument, exogenous variable
#' @param equal logical whether the misclassification probabilities do not depend on instrument
#'
moments <- function(par, data, equal) {

  # the number of the elements V takes
  K <- max(data$V)

  if (equal == FALSE) {

    # parameters
    beta <- par[1]
    delta_p <- par[2]
    r <- par[3]
    tau0 <- par[4]
    tau1 <- par[5]
    m00 <- par[6]
    m01 <- par[7]
    m10 <- par[8]
    m11 <- par[9]
    j <- 0
    for (j in 1:K) {
      assign(sprintf("p0%d", j), par[9 + j])
      assign(sprintf("p1%d", j), par[9 + K + j])
    }

    # functions for moments
    f1 <- beta - (data$Y * data$Z / r - data$Y * (1 - data$Z) / (1 - r)) / delta_p
    f2 <- delta_p - ( (data$D * data$Z / r - m01) / (1 - m01 - m11) - (data$D * (1 - data$Z) / (1 - r) - m00) / (1 - m00 - m10) )
    f3 <- r - data$Z
    j <- 0
    for (j in 1:K) {
      I0v <- data[[sprintf("I0%d", j)]]
      I1v <- data[[sprintf("I1%d", j)]]
      p0v <- get(sprintf("p0%d", j))
      p1v <- get(sprintf("p1%d", j))
      assign(sprintf("f%d", 3 + j), (m00 + (1 - m00 - m10) * p0v - data$D) * I0v)
      assign(sprintf("f%d", 3 + K + j), (m01 + (1 - m01 - m11) * p1v - data$D) * I1v)
      assign(sprintf("f%d", 3 + 2 * K + j), (tau0 + (data$Y * data$D - (1 - m10) * p0v * tau0) / (m00 + (1 - m00 - m10) * p0v) - (data$Y * (1 - data$D) + (1 - m00) * (1 - p0v) * tau0) / (1 - (m00 + (1 - m00 - m10) * p0v))) * I0v)
      assign(sprintf("f%d", 3 + 3 * K + j), (tau1 + (data$Y * data$D - (1 - m11) * p1v * tau1) / (m01 + (1 - m01 - m11) * p1v) - (data$Y * (1 - data$D) + (1 - m01) * (1 - p1v) * tau1) / (1 - (m01 + (1 - m01 - m11) * p1v))) * I1v)
    }

    # return functions for moments
    f <- NULL
    j <- 0
    for (j in 1:(3 + 4 * K)) {
      f <- cbind(f, get(sprintf("f%d", j)))
    }

    return(f)

  } else if (equal == TRUE) {

    # parameters
    beta <- par[1]
    delta_p <- par[2]
    r <- par[3]
    tau0 <- par[4]
    tau1 <- par[5]
    m0 <- par[6]
    m1 <- par[7]
    j <- 0
    for (j in 1:K) {
      assign(sprintf("p0%d", j), par[7 + j])
      assign(sprintf("p1%d", j), par[7 + K + j])
    }
    # functions for moments
    f1 <- beta - (data$Y * data$Z / r - data$Y * (1 - data$Z) / (1 - r)) / delta_p
    f2 <- delta_p - (data$D * data$Z / r - data$D * (1 - data$Z) / (1 - r)) / (1 - m0 - m1)
    f3 <- r - data$Z
    j <- 0
    for (j in 1:K) {
      I0v <- data[[sprintf("I0%d", j)]]
      I1v <- data[[sprintf("I1%d", j)]]
      p0v <- get(sprintf("p0%d", j))
      p1v <- get(sprintf("p1%d", j))
      assign(sprintf("f%d", 3 + j), (m0 + (1 - m0 - m1) * p0v - data$D) * I0v)
      assign(sprintf("f%d", 3 + K + j), (m0 + (1 - m0 - m1) * p1v - data$D) * I1v)
      assign(sprintf("f%d", 3 + 2 * K + j), (tau0 + (data$Y * data$D - (1 - m1) * p0v * tau0) / (m0 + (1 - m0 - m1) * p0v) - (data$Y * (1 - data$D) + (1 - m0) * (1 - p0v) * tau0) / (1 - (m0 + (1 - m0 - m1) * p0v))) * I0v)
      assign(sprintf("f%d", 3 + 3 * K + j), (tau1 + (data$Y * data$D - (1 - m1) * p1v * tau1) / (m0 + (1 - m0 - m1) * p1v) - (data$Y * (1 - data$D) + (1 - m0) * (1 - p1v) * tau1) / (1 - (m0 + (1 - m0 - m1) * p1v))) * I1v)
    }

    # return functions for moments
    f <- NULL
    j < - 0
    for (j in 1:(3 + 4 * K)) {
      f <- cbind(f, get(sprintf("f%d", j)))
    }

    return(f)
  }
}


R/auxiliary.R
#' adding indicators for exogenous variable
#'
#' \code{addindicators} is an auxiliary function for preparation to implement the GMM inference on your own.
#' It returns a data.frame that contains vectors of the outcome, the treatment, the instrument, and the exogenous variable in addition to vectors of indicators for the exogenous variable.
#' The defined indicator is such as I00 = 1(Z = 0, V = 0).
#' See the package vignette via `vignette("LATEerror")` for more details.
#'
#' @param Y vector of the outcome
#' @param D vector of the treatment
#' @param Z vector of the instrument
#' @param V vector of the exogenous variable
#'
#' @return dataframe with indicators for the exogenous variable
#'
#' @export
#'
addindicators <- function(Y, D, Z, V) {

  # make dataframe
  data <- data.frame(Y = Y, D = D, Z = Z, V = V)

  # the number of the elements that V takes
  K <- max(data$V)

  # add indicators
  for (j in 1:K) {
    varname0 <- sprintf("I0%d", j)
    data[[varname0]] <- (1 - data$Z) * ifelse(data$V == j, 1, 0)
    varname1 <- sprintf("I1%d", j)
    data[[varname1]] <- data$Z * ifelse(data$V == j, 1, 0)
  }

  return(data)

}



#' returning the names and the numbers of the parameters
#'
#' @param equal logical whether misclassification probabilities do not depend on Z
#' @param K the number of the elements that V takes
#'
number_name_par <- function(equal, K) {

  if (equal == FALSE) {

    # the number of the parameters
    number_par <- 2 * K + 9

    # the name of the parameters
    name_par <- c("LATE", "first-stage", "E(Z)", "tau0*", "tau1*", "m00", "m01", "m10", "m11")
    for (j1 in 1:K) {
      name_par <- c(name_par, sprintf("p0%d*", j1))
    }
    for (j2 in 1:K) {
      name_par <- c(name_par, sprintf("p1%d*", j2))
    }

  } else if (equal == TRUE) {

    # the number of the parameters
    number_par <- 2 * K + 7

    # the name of the parameters
    name_par <- c("LATE", "first-stage", "E(Z)", "tau0*", "tau1*", "m0", "m1")
    for (j1 in 1:K){
      name_par <- c(name_par, sprintf("p0%d*", j1))
    }
    for (j2 in 1:K){
      name_par <- c(name_par, sprintf("p1%d*", j2))
    }

  }

  # result
  result <- list(number_par = number_par, name_par = name_par)

  return(result)

}


R/inference.R
#' computing standard errors and confidence intervals
#'
#' \code{inference} is an auxiliary function to implement the GMM inference on your own.
#' It returns estimates, standard errors, and confidence intervals
#' for the parameters in the GMM estimation.
#' See the package vignette via `vignette("LATEerror")` for more details.
#'
#' @param par vector of parameter values
#' @param data dataframe that contains Y, D, Z, and V with indicators
#' @param weight weighting matrix for pilot GMM estimation
#' @param equal logical whether misclassification probabilities do not depend on Z
#'
#' @return list that contains the following elements.
#' \item{estimate}{parameter estimates}
#' \item{se}{standard errors}
#' \item{ci}{95 percent confidence intervals}
#'
#' @importFrom MASS ginv
#' @importFrom numDeriv jacobian
#' @importFrom stats qnorm
#'
#' @export
#'
inference <- function(par, data, weight, equal) {

  # sample size
  n <- length(data$Y)

  # the number of the elements in the support of V
  K <- max(data$V)

  # the number of the parameters and the names of the parameters
  temp <- number_name_par(equal = equal, K = K)
  number_par <- temp$number_par
  name_par <- temp$name_par

  # jacobian of the moments
  G <- jacobian(mean_moments, par, data = data, equal = equal)

  # variance-covariance matrix of the moments
  var_moments_estimate <- var_moments(par = par, data = data, equal = equal)

  # computing inverse matrix
  if (qr(t(G) %*% weight %*% G)$rank < number_par) {

    covariance <- ginv(t(G) %*% weight %*% G) %*%
      t(G) %*% weight %*% var_moments_estimate %*% weight %*% G %*%
      ginv(t(G) %*% weight %*% G) / n

    warning("The asymptotic variance-covariance matrix estimation produces a singular matrix.
            The generalized inverse is used.")

  } else {

    covariance <- solve(t(G) %*% weight %*% G) %*%
      t(G) %*% weight %*% var_moments_estimate %*% weight %*% G %*%
      solve(t(G) %*% weight %*% G) / n

  }

  # standard errors
  se <- sqrt(diag(covariance))

  # 95% confidence intervals
  ci <- rbind(par + qnorm(0.025) * se, par + qnorm(0.975) * se)

  # names
  names(par) <- name_par
  names(se) <- name_par
  colnames(ci) <- name_par
  rownames(ci) <- c("95% CI lower", "95% CI upper")

  # results
  result <- list(estimate = par, se = se, ci = ci)

  return(result)

}



#' conducting over-identification test (J test)
#'
#' @param par GMM estimates
#' @param data dataframe containing the outcome, treatment, instrument, exogenous variable with indicators
#' @param equal logical whether the misclassification probabilities do not depend on instrument
#'
#' @return vector that contains a test statistic value and p-value
#'
#' @importFrom stats pchisq
#'
Jtest <- function(par, data, equal) {

  # sample size
  n <- length(data$Y)

  # the number of the elements in the support of V
  K <- max(data$V)

  # the number of the moments
  number_moment <- 3 + 4 * K

  # the number of the parameters and the names of the parameters
  temp <- number_name_par(equal = equal, K = K)
  number_par <- temp$number_par
  name_par <- temp$name_par

  # the means of the moments
  S <- mean_moments(par = par, data = data, equal = equal)

  # over-identification test statistic
  var_moments_estimate <- var_moments(par = par, data = data, equal = equal)
  over_test <- n * S %*% solve(var_moments_estimate) %*% S

  # p value
  df <- number_moment - number_par
  over_test_p <- 1 - pchisq(q = over_test, df = df)

  # results
  overidentification <- c(over_test, over_test_p)
  names(overidentification) <- c("test statistic", "p-value")

  return(overidentification)

}



#' Estimating variance-covariance matrix of the moments
#'
#' @param par GMM estimates
#' @param data dataframe containing the outcome, treatment, instrument, exogenous variable
#' @param equal logical whether the misclassification probabilities do not depend on instrument
#'
var_moments <- function(par, data, equal) {

  # the number of the elements V takes
  K <- max(data$V)

  # the number of the moments
  number_moment <- 3 + 4 * K

  # the values of the moments
  f <- moments(par = par, data = data, equal = equal)

  # variance-covariance matrix estimation
  hat_gamma <- NULL
  for (j in 1:number_moment) {
    hat_gamma <- rbind(hat_gamma, colMeans(f * f[, j]))
  }

  return(hat_gamma)

}


inst/example/example.R
library(LATEerror)
library(dplyr)

# simulated data----------------------------------------------------------------
n <- 2000
set.seed(102)
mydata <- data.frame(U1 = rnorm(n, 0, 1), U2 = rnorm(n, 0, 1))
mydata <- mydata %>%
  dplyr::mutate(Z = rbinom(n, size = 1, prob = 0.5)) %>%
  dplyr::mutate(V = rbinom(n, size = 1, prob = 0.5) + 1) %>%
  dplyr::mutate(Ds = ifelse(( -2 + Z + V - U1 >= 0), 1, 0)) %>%
  dplyr::mutate(D = Ds * rbinom(n, size = 1, prob = 0.75) + (1 - Ds) * rbinom(n, size = 1, prob = 0.25)) %>%
  dplyr::mutate(Y = 1 + Ds + U2)

# GMM estimation----------------------------------------------------------------
set.seed(102)
gmmest <- LATEerror(Y = mydata$Y, D = mydata$D, Z = mydata$Z, V = mydata$V)

# pilot GMM estimation result---------------------------------------------------
# estimates
gmmest$pilot$estimate

# standard errors
gmmest$pilot$se

# 95% confidence intervals
gmmest$pilot$ci

# optimal GMM estimation result-------------------------------------------------
# estimates
gmmest$optim$estimate

# standard errors
gmmest$optim$se

# 95% confidence intervals
gmmest$optim$ci

# over-identification test
gmmest$optim$overidentification


R/LATEerror.R
#' GMM inference to estimate the LATE for a mismeasured binary treatment
#'
#' \code{LATEerror} implements the GMM inference to correct a measurement error
#' problem due to a mismeasured binary treatment in the LATE inference
#' developed in Yanagi (2017).
#' See the package vignette via \code{vignette("LATEerror")} for more details.
#'
#' @param Y vector of the outcome
#' @param D vector of the treatment
#' @param Z vector of the instrument
#' @param V vector of the exogenous variable
#' @param weight weighting matrix for the pilot GMM estimation
#' @param optimal logical whether implementing the optimal GMM estimation
#' @param equal logical whether the misclassification probabilities do not depend on the instrument
#' @param lower vector of lower bounds of the parameters
#' @param upper vector of upper bounds of the parameters
#' @param control list of the control parameters for \code{\link[DEoptim]{DEoptim}}
#'
#' @return list that contains the two lists: (i) \code{pilot}, (ii) \code{optimal}.\cr
#' \code{pilot} contains the following elements.
#' \item{estimate}{vector of parameter estimates}
#' \item{se}{vector of standard errors}
#' \item{ci}{matrix that contains 95 percent confidence intervals}
#' \item{optim}{list that contains the results of the optimization via \code{DEoptim}}
#' \item{weight}{weighting matrix for the pilot GMM estimation}
#' \code{optimal} contains the following elements.
#' \item{estimate}{vector of parameter estimates}
#' \item{se}{vector of standard errors}
#' \item{ci}{matrix that contains 95 percent confidence intervals}
#' \item{overidentification}{vector that contains a value of the test statistic and p-value for the over-identification test}
#' \item{weight}{weighting matrix for the optimal GMM estimation}
#' \item{optim}{list that contains the results of the optimization via \code{DEoptim}}
#'
#' @seealso \code{\link[DEoptim]{DEoptim}} for the elements in \code{control}
#'
#' @importFrom DEoptim DEoptim
#'
#' @export
#'
LATEerror <- function(Y, D, Z, V, weight = NULL, optimal = FALSE, equal = TRUE, lower = NULL, upper = NULL, control = NULL) {

  #-----------------------------------------------------------------------------
  # preparation
  #-----------------------------------------------------------------------------

  # sample size
  n <- length(Y)

  # stop with error messages
  if (n != length(D) || n != length(Z) || n != length(V)) {
    stop(message = "The lengths of Y, D, Z, and V must be the same.")
  }
  if (min(D) != 0 || max(D) != 1 || sum(D == 0) == 0 || sum(D == 1) == 0) {
    stop(message = "D must take two values of 0 and 1.")
  }
  if (min(Z) != 0 || max(Z) != 1 || sum(Z == 0) == 0 || sum(Z == 1) == 0) {
    stop(message = "Z must take two values of 0 and 1.")
  }

  # the number of the elements in the support of V
  K <- max(V)

  # the number of the parameters and the names of the parameters
  number_name <- number_name_par(equal = equal, K = K)
  number_par <- number_name$number_par
  name_par <- number_name$name_par

  # the number of the moments
  number_moment <- 3 + 4 * K

  # stop with error messages
  for (j in 1:K) {
    if (min(V) <= 0 || sum(V == j) == 0) {
      stop(message = "V must take K consecutive discrete values (1, 2, ..., K).")
    }
  }
  if (K <= 2 && equal == FALSE) {
    stop(message = "If 'equal = FALSE', V must take at least three values.")
  }
  if (!is.null(weight)) {
    if (!is.matrix(weight)) {
      stop(message = "'weight' must be NULL or a symmetric matrix.")
    } else if (ncol(weight) != number_moment && nrow(weight) != number_moment) {
      stop(message = "'weight' must be NULL or a symmetric matrix.")
    } else if (t(weight) != weight) {
      stop(message = "'weight' must be NULL or a symmetric matrix.")
    }
  }
  if (!is.logical(optimal)) {
    stop(message = "'optimal' must be logical.")
  }

  # making data frame with the indicators
  data <- addindicators(Y = Y, D = D, Z = Z, V = V)

  # lower and upper bounds for optimization
  bound <- max(Y) - min(Y)
  if (is.null(lower)) {
    if (equal == FALSE) {
      lower <- c(-bound, 0.001, 0.001, -bound, -bound, rep(0.001, 4), rep(0.001, 2 * K))
    } else if (equal == TRUE){
      lower <- c(-bound, 0.001, 0.001, -bound, -bound, rep(0.001, 2), rep(0.001, 2 * K))
    }
  } else {
    if (!is.vector(lower)) {
      stop(message = "The lengths of 'lower' and 'upper' must be identical to that of the parameters.")
    } else if (length(lower) != number_par) {
      stop(message = "The lengths of 'lower' and 'upper' must be identical to that of the parameters.")
    } else if (equal == FALSE) {
      if (lower[6] + lower[7] >= 1 || lower[8] + lower[9] >= 1) {
        stop(message = "The sum of the misclassification probabilities must be smaller than one.")
      }
    } else if (equal == TRUE){
      if (lower[6] + lower[7] >= 1) {
        stop(message = "The sum of the misclassification probabilities must be smaller than one.")
      }
    }
  }
  if (is.null(upper)) {
    if (equal == FALSE) {
      upper <- c(bound, 0.999, 0.999, bound, bound, rep(0.499, 4), rep(0.999, 2 * K))
    } else if (equal == TRUE) {
      upper <- c(bound, 0.999, 0.999, bound, bound, rep(0.499, 2), rep(0.999, 2 * K))
    }
  } else {
    if (!is.vector(upper)) {
      stop(message = "The lengths of 'lower' and 'upper' must be identical to that of the parameters.")
    } else if (length(upper) != number_par) {
      stop(message = "The lengths of 'lower' and 'upper' must be identical to that of the parameters.")
    } else if (equal == FALSE) {
      if (upper[6] + upper[7] >= 1 || upper[8] + upper[9] >= 1) {
        stop(message = "The sum of the misclassification probabilities must be smaller than one.")
      }
    } else if (equal == TRUE) {
      if (upper[6] + upper[7] >= 1) {
        stop(message = "The sum of the misclassification probabilities must be smaller than one.")
      }
    }
  }

  # weight matrix for the pilot estimation
  if (is.null(weight)) {
    weight1 <- diag(number_moment)
  } else {
    weight1 <- weight
  }

  # control for DE algorithm
  if (!is.list(control) && !is.null(control)) {
    stop(message = "'control' must be correctly specified.")
  }
  if (is.null(control$NP)) {
    control$NP <- 10 * number_par
  }
  if (is.null(control$itermax)) {
    control$itermax <- 10000
  }
  if (is.null(control$steptol)) {
    control$steptol <- 500
  }
  if (is.null(control$reltol)) {
    control$reltol <- 1e-12
  }
  if (is.null(control$trace)) {
    control$trace <- 100
  }
  if (is.null(control$CR)) {
    control$CR <- 0.5
  }
  if (is.null(control$F)){
    control$F <- 0.8
  }
  if (is.null(control$c)){
    control$c <- 0.05
  }
  if (is.null(control$strategy)){
    control$strategy <- 2
  }
  if (is.null(control$parallelType)) {
    control$parallelType <- 1
  }

  #-----------------------------------------------------------------------------
  # pilot estimation
  #-----------------------------------------------------------------------------

  # optimization
  estDE1 <- DEoptim(fn = criterion, lower = lower, upper = upper, data = data, equal = equal, weight = weight1, control = control)

  # pilot estimates
  estimate1 <- estDE1$optim$bestmem

  # standard errors and confidence intervals
  inference1 <- inference(par = estimate1, data = data, weight = weight1, equal = equal)

  # pilot estimation result
  estimation1 <- list(estimate = inference1$estimate, se = inference1$se, ci = inference1$ci, optim = estDE1$optim, weight = weight1)

  if (optimal == FALSE) {
    result <- list(pilot = estimation1)
    return(result)
  }

  #-----------------------------------------------------------------------------
  # optimal estimation
  #-----------------------------------------------------------------------------

  # optimal weighting matrix
  var_moments_estimate1 <- var_moments(par = estimate1, data = data, equal = equal)
  weight2 <- solve(var_moments_estimate1)

  # optimization
  estDE2 <- DEoptim(fn = criterion, lower = lower, upper = upper, data = data, weight = weight2, equal = equal, control = control)

  # optimal estimates
  estimate2 <- estDE2$optim$bestmem

  # standard errors for optimal estimation
  inference2 <- inference(par = estimate2, data = data, weight = weight2, equal = equal)

  # over-identication test
  if (number_moment > number_par) {

    overidentification <- Jtest(par = estimate2, data = data, equal = equal)

  } else {

    overidentification <- NULL

  }

  # optimal estimation result
  estimation2 <- list(estimate = inference2$estimate, se = inference2$se,
                      ci = inference2$ci, overidentification = overidentification,
                      weight = weight2, optim = estDE2$optim)

  result <- list(pilot = estimation1, optimal = estimation2)

  return(result)

}
